core/meta/object-descriptor.js

```diff
@@ -357,35 +358,121 @@ var ObjectDescriptor = exports.ObjectDescriptor = Montage.specialize( /** @lends
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    } else if (!this._propertyDescriptorsTable.has(descriptor.name)) {
+                        this._propertyDescriptors.push(descriptor);
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    }
+                    descriptor._owner = descriptor._owner || this;
+                }
             }
+        }
+    },
 
+    _preparePropertyDescriptorsCache: {
+        value: function () {
+            var ownDescriptors = this._ownPropertyDescriptors,
+                isReady = true,
+                descriptor, i, n;
+            
+            if (!this._propertyDescriptors)  {
+                
+                for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
+                    descriptor = ownDescriptors[i];
+                    isReady = !!(descriptor && descriptor.name);
+                }
+    
+                if (isReady) {
+                    this._propertyDescriptors = [];
+                    this._propertyDescriptorsTable.clear();
+                    for (i = 0, n = ownDescriptors.length; i < n && isReady; ++i) {
+                        descriptor = ownDescriptors[i];
+                        descriptor._owner = this;
+                        this._propertyDescriptors.push(descriptor);
+                        this._propertyDescriptorsTable.set(descriptor.name,  descriptor);
+                    }
+                    this.addRangeAtPathChangeListener("_ownPropertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
+                    this.addRangeAtPathChangeListener("parent.propertyDescriptors", this, "_handlePropertyDescriptorsRangeChange");
+                }
+            }
         }
     },
 
     /**
-     * @returns {Array.<PropertyDescriptor>}
+     * PropertyDescriptors for this object descriptor, not including those 
+     * provided by this.parent
+     *
+     * @property {Array<PropertyDescriptor>} value
+     */
+    _ownPropertyDescriptors: {
+        value: null
+    },
+
+    _propertyDescriptors: {
+        value: null
+    },
+
+    /**
+     * PropertyDescriptors associated to this object descriptor, including those 
+     * provided by this.parent.
+     *
+     * @property {Array<PropertyDescriptor>} value
      */
     propertyDescriptors: {
         get: function () {
-            var propertyDescriptors = [];
-            propertyDescriptors = propertyDescriptors.concat(this._propertyDescriptors);
-            if (this.parent) {
-                propertyDescriptors = propertyDescriptors.concat(this.parent.propertyDescriptors);
+            this._preparePropertyDescriptorsCache();
+            return this._propertyDescriptors || [];
+        }
+    },
+
+    _validParentPropertyDescriptors: {
+        value: function () {
+            var parentPropertyDescriptors = this.parent.propertyDescriptors.slice(), 
```

Comment:
I don't see why we're making a copy (slice) of this.parent.propertyDescriptors here, we're looking on content and adding objects to something else

Reply:
Hmm.. Me neither. I would hope that I would not add that unnecessarily, but I cannot see a reason on first inspection. I will remove the slice and run the tests. If all goes well I'll leave it off.